File: lib/containers/splitter_dock_container.dart
part of dock_spawn;

abstract class SplitterDockContainer implements IDockContainer {
  String name;
  DockManager dockManager;
  SplitterPanel splitterPanel;
  String containerType;
  bool get stackedVertical;

  int _cachedWidth;
  int _cachedHeight;
  SplitterDockContainer(this.name, this.dockManager, List<IDockContainer> childContainers) {
    splitterPanel = new SplitterPanel(childContainers, stackedVertical);
  }
  
  
  void resize(int _width, int _height) {
//    if (_cachedWidth == _cachedWidth && _cachedHeight == _height) {
//      // No need to resize
//      return;
//    }
    splitterPanel.resize(_width, _height);
    _cachedWidth = _width;
    _cachedHeight = _height;
  }

  int get minimumAllowedChildNodes { return 2; }
  
  void performLayout(List<IDockContainer> childContainers) {
    splitterPanel.performLayout(childContainers);
  }
  
  void setActiveChild(IDockContainer child) {
  }
  
  void destroy() {
    splitterPanel.destroy();
  }

  /**
   * Sets the percentage of space the specified [container] takes in the split panel
   * The percentage is specified in [ratio] and is between 0..1
   */ 
  void setContainerRatio(IDockContainer container, num ratio) {
    splitterPanel.setContainerRatio(container, ratio);
    resize(width, height);
  }
  
  void saveState(Map<String, Object> state) {
    state['width'] = width;
    state['height'] = height;
  }
  
  void loadState(Map<String, Object> state) {
    num _width = state['width'];
    num _height = state['height'];
    resize(_width, _height);
  }
  
  Element get containerElement {
    return splitterPanel.panelElement;
  }

  int get width {
    if (_cachedWidth == null) {
      _cachedWidth = splitterPanel.panelElement.$dom_clientWidth;
    }
    return _cachedWidth;
//    return splitterPanel.panelElement.clientWidth;
  }
  
  int get height {
    if (_cachedHeight == null) {
      _cachedHeight = splitterPanel.panelElement.$dom_clientHeight;
    }
    return _cachedHeight; //splitterPanel.panelElement.clientHeight;
//    return splitterPanel.panelElement.clientHeight;
  }

}

File: js/lib/containers/SplitterDockContainer.js
import "../splitter/SplitterPanel";

dockspawn.SplitterDockContainer = function(name, dockManager, childContainers)
{
    // for prototype inheritance purposes only
    if (arguments.length == 0)
        return;

    this.name = name;
    this.dockManager = dockManager;
    this.splitterPanel = new dockspawn.SplitterPanel(childContainers, this.stackedVertical);
    this.containerElement = this.splitterPanel.panelElement;
    this.minimumAllowedChildNodes = 2;
}

dockspawn.SplitterDockContainer.prototype.resize = function(width, height)
{
//    if (_cachedWidth == _cachedWidth && _cachedHeight == _height) {
//      // No need to resize
//      return;
//    }
    this.splitterPanel.resize(width, height);
    this._cachedWidth = width;
    this._cachedHeight = height;
};

dockspawn.SplitterDockContainer.prototype.performLayout = function(childContainers)
{
    this.splitterPanel.performLayout(childContainers);
};

dockspawn.SplitterDockContainer.prototype.setActiveChild = function(child)
{
};

dockspawn.SplitterDockContainer.prototype.destroy = function()
{
    this.splitterPanel.destroy();
};

/**
 * Sets the percentage of space the specified [container] takes in the split panel
 * The percentage is specified in [ratio] and is between 0..1
 */
dockspawn.SplitterDockContainer.prototype.setContainerRatio = function(container, ratio)
{
    this.splitterPanel.setContainerRatio(container, ratio);
    this.resize(this.width, this.height);
};

dockspawn.SplitterDockContainer.prototype.saveState = function(state)
{
    state.width = this.width;
    state.height = this.height;
};

dockspawn.SplitterDockContainer.prototype.loadState = function(state)
{
    this.resize(state.width, state.height);
};

Object.defineProperty(dockspawn.SplitterDockContainer.prototype, "width", {
    get: function()
    {
        if (this._cachedWidth === undefined)
            this._cachedWidth = this.splitterPanel.panelElement.clientWidth;
        return this._cachedWidth;
    }
});

Object.defineProperty(dockspawn.SplitterDockContainer.prototype, "height", {
    get: function()
    {
        if (this._cachedHeight === undefined)
            this._cachedHeight = this.splitterPanel.panelElement.clientHeight;
        return this._cachedHeight;
    }
});
